docs/DesignDocs/ReproducibleBuildUsingNix.md

```diff
@@ -0,0 +1,397 @@
+   ubuntu 20.04 because package versions align best with nixpkgs release-20.03.
+
+ - BUILD_USER, BUILD_USER_ID, BUILD_USER_HOME need to be set together.
+   Reproducibility requires a standard build user, since the tar directory entries in the deb package include 
+   user and group information.  We default to "azureuser" "1000" "/home/azureuser"
+
+The script "build-oe-nix-image.sh" is a thin wrapper. It takes no arguments, and just passes the needed arguments
+to docker build.  If it needs to be modified, a new script should be created.
+
+Dockerfile.nix prepares the build image by:
+
+- adding via apt minimum packages for the outer image. This includes a text editor (vim) and ELF patch utility (patchelf).
+
+- sets up nit configuration.
+
+- diables apt.
+
+- creates the build user and directories, groups etc.
+
+- creates a nix expression for building openenclave via (shell.nix).
+
+- installs nix as the build user
+
+- calls the prep-nix-build.sh script which sets up the git clone of nix pkgs and preinstalls the package
+  derivations needed for building openenclave into the nix store. The search process for dependencies in
+  nix is to refer to the nix store first, before looking at nixpkgs or contacting the cache if enabled. While
+  we cannot select package versions from the expression, but must depend on the nixpkgs repo to provide the
+  correct versions, we can prime the nix store to provide the exact derivation we wish before referring to the repo.
+
+  These derivations have been manually chosen.  This also allows us to set package priorities which is required
+  between packages that provide overlapping contents, for example clang_7 and llvm_7.
+
+- adds a build.env file which contains environment values to control the build.  This is not part of the 
+  - REV is the git revision (commit or tag) to pull for the build.
+  - SHA is the sha256 hash of the tar file downloaded from git.
+  - DO_CHECK enables ctests as part of the build.
+  - DEB_SHA specifies the expected result sha25sum for the .deb file produced. An error will be thrown if the 
+    sha is incorrect. If not set, the sha is not checked.
+
+- adds needed files for ctest to the .nix_libs directory, which will be referred to by ctest. Nix expressions
+  cannot refer to libraries outside the minimum environment and explicit inputs, so these libraries are needed
+  to use sgx (and eventually TrustZone) facilties.
+
+- sets the build user as the user. Among other things, this means the build image when run cannot be 
+  modified by the addition of apt packages.
+
+The resulting image is named openenclave-build. It can (and should) be generated once and pushed into a common 
+repo.
+
+### build_openenclave.sh
+
+Runs the image produced by docker build, with arguments resolved.
+The output of the process is a .deb and a .nar file containing an exported nix store for use in developing 
+openenclave sdk apps using nix.
```

Comment:
Is the instruction to use the OE SDK .deb and other tool chain to build an OE App captured in this design doc or some other doc?

Reply:
Both a directory readme.md and the end of the design doc.

Reply:
I only found "Work Flow" section at the end of the document. My understand is that that section only talks about building the OE SDK, not about how a OE SDK user makes a reproducible build of an OE App. What I was referring to is an equivalent of https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/buildandsign.md, but with instruction to make a reproducible build. Probably also need to use a docker container with specific clang/gcc NIX package?

Reply:
It has that.  
All the steps needed to create a reproducible package are done in the container.  There is no need for user intervention.

Building the build image is optional in the sense that if it was pushed somewhere one could pull it. Other than that, just run the build_openenclave.sh script and a suitable package will show up in the output directory.  In this, it is similar to linux-sgx/linux/reproducibility, which likewise just builds the package.  

Unlike the linux, it is quite practical to shift to a new version/tag and I probably should discuss that more.